Standards/scs-0214-v1-k8s-node-distribution.md

```diff
@@ -0,0 +1,91 @@
+due to wear and tear of hardware, software misconfigurations, external problems or
+user errors. Whichever was the source of such an outage, it always means down-time for
+operations and users and possible even data loss.
+Therefore, a Kubernetes cluster in a productive environment should be distributed over
+multiple "failure zones" in order to provide fault-tolerance and high availability.
+This is especially important for the control plane of the cluster, since it contains the
+state of the whole cluster. A failure of this component could mean an unrecoverable failure
+of the whole cluster.
+
+## Design Considerations
+
+Most design considerations of this standard follow the previously written Decision Record
+"Kubernetes Nodes Anti Affinity" as well as the Kubernetes documents about [High Availability]
+and [Best practices for large clusters].
+
+SCS wishes to prefer distributed, highly-available systems due to their obvious advantages
+like fault-tolerance and data redundancy. But it also understands the costs and overhead
+for the providers associated with this effort, since the infrastructure needs to have
+hardware which will just be used to provide fail-over safety or duplication.
+
+The document [Best practices for large clusters] describes the concept of a failure zone.
+This term isn't defined any further, but can in this context be described as a number of
+physical (computing) machines in such a vicinity to each other (either through physical
+or logical interconnection in some way), that specific problems inside this zone would put
+all these machines at risk of failure/shutdown. It is therefore necessary for important
+data or services to not be present just on one failure zone.
+How such a failure zone should be defined is dependent on the risk model of the service/data
+and its owner as well as the capabilities of the provider. Zones could be set from things
+like single machines or racks up to whole datacenters or even regions, which could be
+coupled by things like electrical grids. They're therefore purely logical entities, which
+shouldn't be defined further in this document.
+
+## Decision
+
+This standard formulates the requirement for the distribution of Kubernetes nodes in order
+to provide a fault-tolerant and available Kubernetes cluster infrastructure.
+Since some providers only have small environments to work with and therefore couldn't
+comply with this standard, it will be treated as a RECOMMENDED standard, where providers
+can OPT-OUT.
+
+If the standard is used by a provider, the following decisions are binding and valid:
+
+- The control plane nodes SHOULD be distributed over multiple physical machines. Kubernetes
+  provides best-practices on this topic, which are also RECOMMENDED by SCS.
+- At least one control plane instance SHOULD be run in each "failure zone", more are
+  RECOMMENDED in each "failure zone" to provide fault-tolerance for each zone.
```

Comment:
I think that it is important to make this more concrete, even if the linked Kubernetes docs already say that "three or more" machines are required for HA.

For example, if only two "failure zones" are present, hosting one or more etcd members each, this would have no real advantage from a robustness POV because the etcd quorum requires more than half of its members to be online.

So, AFAIU, this standard should mandate at least 3 "failure zones" with at least one "control plane instance" (or etcd member) running inside.
Running more probably should be considered non-mandatory (or "less strictly recommended").

Maybe it should also be considered to enable implementers to split their etcd operations from apiserver/scheduler/controller-manager/... operation, but I understand that this might make things more complicated, while covering an unusual case.

Reply:
I agree with the part about the failure zones; this should be integrated in the standard.

What I'm not sure about is the separation of the etcd operation. I understand where you coming from here, but this adds quite a bit more complexity and also sets even stricter requirements for the providers.
I would like to have a broader discussion about this (especially with the providers) and then maybe put this part in the second version of this standard (or even a separate one, if it comes to this).